File: DSA_python/DSA01009 XÂU AB ĐẶC BIỆT.py
def next(A, n):
    for i in range(n-1, -1, -1):
        if A[i] == 'A':
            A[i] = 'B'
            break
        A[i] = 'A'

def maximum(n):
    total = 1
    for i in range(1, n+1):
        total *= 2
    return total

def submit():
    n, k = map(int, input().split())
    A = ['A']*n
    ans = list()

    for i in range(maximum(n)):
        check = False
        count = 0
        for j in A:
            if j == 'A':
                count+=1
                check = True
            elif j == 'B' and check == True:
                if count == k:
                    break
                else:
                    count = 0
                    check = False
            
        if count == k:
            ans.append(A[:])
        next(A, n)
    
    print(len(ans))
    for i in ans:
        print("".join(map(str, i)))



submit()

File: DSA/DSA01009 XÂU AB ĐẶC BIỆT.cpp
#include <bits/stdc++.h>
using namespace std;

void next(vector<char> &A, int n)
{
    for (int i = n - 1; i >= 0; i--)
    {
        if (A[i] == 'A')
        {
            A[i] = 'B';
            break;
        }
        A[i] = 'A';
    }
}

long long maximum(int n)
{
    long long total = 1;
    for (int i = 1; i <= n; i++)
    {
        total *= 2;
    }
    return total;
}

void submit()
{
    int n, k;
    cin >> n >> k;

    vector<char> A(n, 'A');
    vector<vector<char>> ans;

    for (int i = 0; i < maximum(n); i++)
    {
        bool check = false;
        int count = 0;
        for (auto &&j : A)
        {
            if (j == 'A')
            {
                count++;
                check = true;
            }
            else if (j == 'B' && check)
            {
                if (count == k)
                {
                    break;
                }
                else if (count != k)
                {
                    count = 0;
                    check = false;
                }
            }
        }
        
        if (count == k)
        {
            ans.push_back(A);
        }
        next(A, n);
    }
    
    cout << ans.size() << endl;
    for (auto &&i : ans)
    {
        for (auto &&j : i)
        {
            cout << j;
        }
        cout << endl;
    }
}

int main()
{
    ios::sync_with_stdio(false);                                    // Speed up input/output
    cin.tie(nullptr);
    cout.tie(nullptr);

    
    submit();

    return 0;
}